### PR TITLE
[Perf][DSpark] lossless improve DSpark throughput by W2 matrix topk pruned

### PR DESCRIPTION
**Pull request by mugglewei97**
## Purpose
According to the work and research of DFlash2, the vast majority of the probability mass at each position in the draft model is concentrated on a small set of top-k tokens.   

For the Markov matrix B = W1 @ W2 in DSpark's speculative decoding process, W2 has a shape of [rank, full vocab]. Due to the long-tailed distribution of the probability mass, many draft tokens have an pretty low probability of being accepted by the target model, making it unnecessary to compute a full-vocabulary(151936, Qwen3-4B for example)  Markov state transition matrix. 

Therefore, the W2 matrix can be pruned to [rank, top-k] by indexing only the  top-k rows. While introducing almost no reduction in accepted length, this reduces both the floating-point operations and memory access of the W2 matrix, thereby further improving DSpark's output throughput in a training-free and plug-and-play manner. 

This method is fully backward-compatible and has zero impact on the existing DSpark code.

This method has been benchmarked and consistently achieves performance gains (**brings about 5%-8% output throughput**) across all hyper parameter configurations.

Guaranteed by target model's verify and sampling, the output distribution is lossless from the aspect of spec decoding.

## Core Math Mechanism
considering the markov matrix of dspark is 
`B = W1 @ W2`
where W1 shape is [full vocab, rank] and W2 shape is [rank, full vocab], B shape is [full vocab, full vocab] which demonstrate for each pre token, its probability transfer bias  of current candidate token. rank is default 256 

during actual spec decoding, since pre token is confirmed, it could be 
`B = W1[t-1] @ W2`
W1[t] shape would be [1, rank], B shape would be [1, full vocab]

let 
` Ut = top(logits, k) ∪ top(markov bias, k)`
where logits means backbone output logits, markov bias means 
so 
`B = W[t-1] @ W2[Ut]`
W2[Ut] shape is much small than W2

considering throughput depends on `accept_len/(drafting + verifying)`
decreasing drafting would increasing throughput

## Test Plan
on 8*hopper, sweeping across different combinations of base logits top-k and Markov bias top-k values (1, 8, 16, 32, 64, 128) on the GSM8K and C-Eval benchmarks
we found that it almost consistently yields positive throughput gains compared to the baseline.
```bash
vllm serve Qwen/Qwen3-4B \
  --no-enable-flashinfer-autotune \
  --max-num-seqs 128 \
  --data-parallel-size 8 \
  --no-enable-prefix-caching \
  --speculative-config '{
        "method":"dspark",
        "model":"/personal/checkpoints/speculators/dspark_qwen3_4b", 
        "num_speculative_tokens": 8, 
        "markov_topk": 16, 
        "markov_bias_topk": 16
    }'
```
```bash
vllm bench serve \
  --base-url http://localhost:8000 \
  --model Qwen/Qwen3-4B \
  --dataset-name custom \
  --dataset-path /personal/data/ceval/ceval-1040-prompt.jsonl \
  --num-prompts 1040 \
  --temperature 0 \
  --custom-output-len 3072 \
  --no-oversample \
  --max-concurrency 32
```
dspark ckpt can be download here：https://huggingface.co/AQ-MedAI/Qwen3-4B-DSpark

## Test Result
also we benchmarked on concurrence=1, or model=Qwen3-8B, both gain output throughput improvement, which means this feature can be transferred and generalize
### benchmark on gsm8k dataset, temperature=0
image
### benchmark on gsm8k dataset, temperature=1
image

### benchmark on ceval dataset, temperature=0
image
### benchmark on ceval dataset, temperature=1
image

---
<details>
<summary> Essential Elements of an Effective PR Description Checklist </summary>

- [x] The purpose of the PR, such as "Fix some issue (link existing issues this PR will resolve)".
- [x] The test plan, such as providing test command.
- [x] The test results, such as pasting the results comparison before and after, or e2e results
- [ ] (Optional) The necessary documentation update, such as updating `supported_models.md` and `examples` for a new model.
</details>

### CODE REVIEW

**Comment by mugglewei97** [comment]
@youkaichao  @mgoin hello, i develop a new feature of lossless dspark throughput improve, would you plz help me to add ready label so CI can run? 
it is compatible has been fully benchmarked
thank you so much

**Comment by mugglewei97** [comment]
hi @benchislett 
Could you take a look at this PR? It builds on PR [#49969](https://github.com/vllm-project/vllm/pull/49969), which you're familiar with, and includes some performance optimizations.

algorithm aspect：
1）use markov bias top-m as an union with base logits top-k，less row for more accepted candidate

engineer aspect for higher performance and throughput：
1）triton kernel fuse 
2）pre compute w1[t]@w2[topk] as cache instead of compute during online decoding 
3）flashinfer topk，O（V），instead of pytorch.topk，O（Vlogk）

**Comment by eldarkurtic** [comment]
@mugglewei97 could you please run some evals with high entropy data (e.g. [from SPEED bench](https://huggingface.co/datasets/nvidia/SPEED-Bench#:~:text=A%20Throughput%20data,of%20speculative%20decoding.))?

**Comment by mugglewei97** [comment]
> @mugglewei97 could you please run some evals with high entropy data (e.g. [from SPEED bench](https://huggingface.co/datasets/nvidia/SPEED-Bench#:~:text=A%20Throughput%20data,of%20speculative%20decoding.))?

@eldarkurtic
Sure no problem
Here i benchmark SPEED bench 16K in temperature=0
image
with highest 5.0% throughput improvement, if using recommend markov topk=32/markov bias topk=32, would gain 4.2% throughput

also i benchmark SPEED bench 8K in temperature=0
image
also gain 5.3% throughput improvement using recommend markov topk=32/markov bias topk=32
